hibernate-envers/src/test/java/org/hibernate/envers/test/integration/reventity/RevisionNumberOverflowTest.java

```diff
@@ -0,0 +1,154 @@
+import org.hibernate.envers.RevisionEntity;
+import org.hibernate.envers.RevisionNumber;
+import org.hibernate.envers.RevisionTimestamp;
+import org.hibernate.envers.exception.AuditException;
+import org.hibernate.envers.test.BaseEnversJPAFunctionalTestCase;
+import org.hibernate.envers.test.Priority;
+import org.hibernate.envers.test.entities.StrTestEntity;
+import org.hibernate.id.enhanced.TableGenerator;
+import org.junit.Test;
+
+import org.hibernate.testing.TestForIssue;
+
+import static org.hibernate.testing.junit4.ExtraAssertions.assertTyping;
+import static org.hibernate.testing.transaction.TransactionUtil.doInJPA;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * This test checks that when revision number overflow occurs an {@link AuditException} is thrown.
+ *
+ * In order to test this use case, the {@code REVISION_GENERATOR} is explicitly initialized at
+ * {@link Integer.MAX_VALUE} and we attempt to persist two entities that are audited.  The
+ * expectation is that the test should persist the first entity but the second should throw the
+ * desired exception.
+ *
+ * Revision numbers should always be positive values and always increasing, this is due to the
+ * nature of how the {@link org.hibernate.envers.AuditReader} builds audit queries.
+ *
+ * @author Chris Cranford
+ */
+@TestForIssue(jiraKey = "HHH-6615")
+public class RevisionNumberOverflowTest extends BaseEnversJPAFunctionalTestCase {
+	@Override
+	protected Class<?>[] getAnnotatedClasses() {
+		return new Class<?>[] { StrTestEntity.class, CustomCappedRevEntity.class };
+	}
+
+	@Priority(10)
+	@Test
+	public void initData() {
+		// Save entity with maximum possible revision number
+		doInJPA( this::entityManagerFactory, entityManager -> {
+			final StrTestEntity entity = new StrTestEntity( "test1" );
+			entityManager.persist( entity );
+		} );
+
+		// Save entity with overflow revision number
+		try {
+			doInJPA( this::entityManagerFactory, entityManager -> {
+				final StrTestEntity entity = new StrTestEntity( "test2" );
+				entityManager.persist( entity );
+			} );
+		} catch ( Exception e ) {
+			assertRootCause( e, AuditException.class, "Negative revision numbers are not allowed" );
+		}
+	}
+
+	@Test
+	public void testRevisionExpectations() {
+		final StrTestEntity expected = new StrTestEntity( "test1", 1 );
+
+		// Verify there was only one entity instance saved
+		List results = getAuditReader().createQuery().forRevisionsOfEntity( StrTestEntity.class, true, true ).getResultList();
+		assertEquals( 1, results.size() );
+		assertEquals( expected, results.get( 0 ) );
+
+		// Verify entity instance saved has revision Integer.MAX_VALUE
+		assertEquals( expected, getAuditReader().find( StrTestEntity.class, 1, Integer.MAX_VALUE ) );
+	}
+
+	private static void assertRootCause(Exception exception, Class<?> type, String message) {
+		Throwable root = exception;
+		while ( root.getCause() != null ) {
+			root = root.getCause();
+		}
```

Comment:
Should we take into consideration of '**suppressed exception**' (like suppressed in **try-fina**l block)?

Reply:
I don't think that's necessarily needed here given the code flow, wouldn't you think?

Reply:
I am a contributor who doesn't enjoy complete codebase knowledge. Just wanna raise some awareness and I thought it might not hurt.

Reply:
Makes sense, appreciate it.